Vulkan: Feedback loop detection and barriers

riperiperi opened the pull request:
This PR allows the Vulkan backend to detect attachment feedback loops. These are currently used in the following ways:

- Partial use of VK_EXT_attachment_feedback_loop_layout
  - All renderable textures have AttachmentFeedbackLoopBitExt
  - Compile pipelines with Color/DepthStencil feedback loop flags when present
- Support using FragmentBarrier for feedback loops (fixes regressions from https://github.com/Ryujinx/Ryujinx/pull/7012 )

Generally trying to fix issues with RDNA3 GPUs, but I don't have one so it's more of a wild guess.

### API complaints

Feedback loops are very poorly considered by Vulkan and its validation layers, and the addition of the `VK_EXT_attachment_feedback_loop_layout` extension hasn't really helped much. Original specification for Vulkan prohibited attachment feedback loops entirely, simply saying they are entirely disallowed. Software written with other APIs in mind, however, require feedback loops to function properly. Up until recently, the way this issue has been dealt with is for the "general" image layout to unofficially allow feedback loops on desktop GPU drivers. While this isn't to spec, it gives developers access to this capability of the GPU. AMD took the specification very literally and ran - they decided to make the general layout for render targets incompatible with feedback loops.

Feedback loop detection existed in the Vulkan validation layers for a long time, where it was subsequently ignored by everyone writing compatibility layers before the extension was introduced. You'll be shocked to find out that after the extension started being implemented, the validation for feedback loops was actually _entirely removed_ because it wasn't working properly. This leaves no way to check if you're detecting feedback loops properly, and what Vulkan really wants from you, especially in weirder cases like sampling from a depth texture that is read-only. See validation layers issue 8096, I'm not linking it because making anyone related to Vulkan standards read this section would probably count as harassment.

Information on what constitutes a feedback loop appears here:
https://vulkan.lunarg.com/doc/view/1.3.250.1/windows/1.3-extensions/vkspec.html#renderpass-feedbackloop
I only found this a few days ago. I'm not sure when it materialized but it is useful, even if it doesn't mention what source stages and access a "pipeline barrier" for an attachment loop would actually have. It's also kind of weird around what counts as a "read-only" access, since you can disable depth write but still give the render pass a store action, and previous draws in the pass can also write when the current draw doesn't, etc.

There is no information anywhere about `VK_DEPENDENCY_FEEDBACK_LOOP_BIT_EXT`. The validation layers have no meaningful checks relating to it. DXVK doesn't even use this flag. What does it do, where do you define it? I have no idea, so I'm simply not using it.

Vulkan specification states that "VK_IMAGE_LAYOUT_GENERAL supports all types of device access.". For this reason, I've implemented parts of the feedback loop extension that tell the GPU when a texture _will be used_ in a feedback loop: usage flags, pipeline flags, extension enable flag. It's possible it also requires the dependency flag on the render pass, but I have no idea since nothing really mentions what it's for. If a layout transition is _required_ to do this, it's nullifying the whole point of the general layout even existing. Usage is random and unpredicatable in Ryujinx, so the general layout is normally our best bet, and the best performing on desktop.

### TODO
- ~~AMD GPUs may need layout transitions for it to properly allow textures to be used in feedback loops.~~
  - The proprietary driver doesn't, RADV does.
- ~~Use dynamic state for feedback loops. The background pipeline will always miss since feedback loop state isn't known on the GPU project.~~
- How is the barrier dependency flag used? (DXVK just ignores it, there's no vulkan validation...)
- Add subpass dependencies to fix validation errors (external dependency for attachment->sample mid-pass?)
- ~~Silk.NET needs to be updated for this extension, so we need to deal with issues finding the MoltenVK binary. It would help if we could submit our own lookup paths.~~
  - Resolved via manually overriding path resolvers. Not pretty but sidesteps the issue for now. We can consider properly setting up loading mvk via libvulkan when we are able to update to the latest version.

### RDNA3 Testing
This PR is draft because I want to know if what I've got right now is enough to prevent the issue seen in https://github.com/Ryujinx/Ryujinx/issues/5306 , on both RADV and the Windows driver.

If it isn't avoided, then I can follow up with a change that switches the layout of images used in feedback loops, but I'd rather not do this as it can cause other issues.

GamerzHell9137 commented:
- Fixes regression on Mario Golf: Super Rush and Splatoon 3 on Nvidia GPUs on Windows (2000/3000/4000 series)

It was tested on RTX 3060 Ti with the 560.94 (latest at the moment) drivers.

### Mario Golf: Super Rush
Before:

![image](https://github.com/user-attachments/assets/2ffc6ae9-03e4-4da5-acc5-f25ac49feb57)

After:

![image](https://github.com/user-attachments/assets/28e52368-5c7c-4e9c-86b2-797c84c6ba78)

### Splatoon 3

Before:

![image](https://github.com/user-attachments/assets/b781fe99-bdae-410d-a926-67fdb9153e30)

After:

![image](https://github.com/user-attachments/assets/0198e108-5b9e-424e-81aa-e8dfb5259cdd)

This change is affecting the graphics in Team Sonic Racing on v1.0.0 where barriers are broken. The issue is less prevalent but still exists.

Before:
![image](https://github.com/user-attachments/assets/49d5a3fd-b2c9-43f9-b021-ab396f36a5f1)

After:

![image](https://github.com/user-attachments/assets/3b0ea977-6aa2-407c-ae17-95a5efc3e1f5)

Sadly TOTK is not affected and still exhibits the same issue

![image](https://github.com/user-attachments/assets/fe0a0b40-d779-466f-87f7-6fc0702095f5)

golivax commented:
> ASUS ROG Ally (Windows), purple lines on the ground seem to be gone. Seems to solve 5306.

Yep, I also don't see the purple lines on the ground anymore. Woohoo!

riperiperi commented:
I've seen reports that it doesn't fix RADV. It looks like it is unable to figure out when feedback loops are active with the GENERAL layout, unlike the Windows driver, which probably violates spec. I imagine the feedback loop usage flags, pipeline state and dependency flags should be used to do this, but the driver doesn't seem to care. It seems tailored to how DXVK works.

This really isn't an appropriate reason to implement layout transtions in an application with unpredictable usage, but I'll see how much I can do.

There might be validation errors and times where the wrong layout is used, because it is too complicated and expensive to update the layout in every location every time it changes, so I'll limit the code to only run on RADV.

BooBerry commented:
Can the Windows portion that fixes it there be used still? Been using it for two days now and it seems to be completely fixed on Windows at least.

riperiperi commented:
AMDVLK would probably work, but I don't expect people to switch to it just to run Ryujinx, or to figure out how to, since RADV is pretty standard on linux.

The other changes still need doing, mostly dynamic state and some investigation into Silk.NET on macos.

BooBerry commented:
Indeed, tried it with AMDVLK and it works fine there too.

On Ubuntu (and likely others like Mint, Debian, etc.) and Fedora it's not hard to switch to AMDVLK. In fact you can have both RADV and AMDVLK installed at the same time and you can switch between them by adding and setting `AMD_VULKAN_ICD=` to either `RADV` or `AMDVLK` in the /etc/environment file.

poudink commented:
SteamOS doesn't include AMDVLK and doesn't give you any simple way of installing it.

Ammar-Sadaoui commented:
I've always wondered why there isn't a built-in setting for switching drivers in Linux. I'm not sure if it would be worth the effort to develop one and submit it as a PR to KDE or COSMIC, only to risk it being rejected

BooBerry commented:
> I've always wondered why there isn't a built-in setting for switching drivers in Linux. I'm not sure if it would be worth the effort to develop one and submit it as a PR to KDE or COSMIC, only to risk it being rejected

Well... it's Linux, nothing's ever that easy. I'm actually surprised installing and switching to AMDVLK on Ubuntu was as easy as it is.

I think someone mentioned the Steam Deck and SteamOS. While it's possible to install and use AMDVLK, SteamOS on the Deck being an immutable OS makes things complicated, forcing the user to enable write and use pacman to install the amdvlk package. This really isn't ideal as it's meant to be immutable and any system updates will wipe any changes made to the system.

IsaacMarovitz commented on src/Ryujinx.Graphics.Vulkan/MoltenVK/MVKInitialization.cs:
```diff
@@ -29,5 +30,20 @@ public static void Initialize()
+        {
+            if (path.EndsWith("libvulkan.dylib"))
+            {
+                path = path[..^"libvulkan.dylib".Length] + "libMoltenVK.dylib";
```
```suggestion
                path.Replace("libvulkan.dylib", "libMoltenVK.dylib");
```

IsaacMarovitz replied:
I guess this has the potential to screw up paths if they contain `libvulkan.dylib` somewhere earlier

riperiperi replied:
It's also less explicit about what I want to do here. It's replacing the filename `libvulkan.dylib`, which only appears at the end. Any extra work done to scan for the match or insert the path isn't necessary.

Nassi88 replied:
how do you fix that. can you please explain it to me for Windows

Nassi88 replied:
I don't understand why I don't have that in yuzu. But I do have it in sudachi and suyu? They are forks. I have it in ryujinx too

Squall-Leonhart replied:
> I don't understand why I don't have that in yuzu. But I do have it in sudachi and suyu? They are forks. I have it in ryujinx too

AMD implemented a application specific workaround that disables Arbitrary DCC format reinterpretation for yuzu specifically

Vinicium commented:
![imagedada](https://github.com/user-attachments/assets/8441bbae-02a5-48c6-90e0-0c5ae76f4c45)
![imaadadge](https://github.com/user-attachments/assets/90b05afc-a31a-4fc8-8a31-b7f94e1a195f)
Worked for me. Rx 7600 driver 24.5.1 i5 12400 W11 23H2

BuyMyMojo commented:
TOTK visual bugs are gone using this PR on my 7800xt with drivers

Lieutenant2014 commented:
TOTK visual bugs are gone on my 7900xtx

creativecoder commented:
Noting that this doesn't seem to fix the issue on Linux. I tested TOTK on Pop!_OS 22.04, Radeon 7800 XT, Mesa 24.0.3

| **Ryulinx 1.1.1379** | **Ryujinx 1.1.0+14c02a9<br/>(build artifact from this PR)** |
| - | - |
| ![Screenshot from 2024-08-26 14-14-40](https://github.com/user-attachments/assets/275f76ca-30a4-4087-8a0a-6eec77924a3c) | ![Screenshot from 2024-08-26 14-23-41](https://github.com/user-attachments/assets/03b61bc3-9ea7-4824-be85-f070de7ba3d4) |

BooBerry commented:
> Noting that this doesn't seem to fix the issue on Linux. I tested TOTK on Pop!_OS 22.04, Radeon 7800 XT, Mesa 24.0.3
> 
> **Ryulinx 1.1.1379**	**Ryujinx 1.1.0+14c02a9(build artifact from this PR)**
> ![Screenshot from 2024-08-26 14-14-40](https://github.com/user-attachments/assets/275f76ca-30a4-4087-8a0a-6eec77924a3c)	![Screenshot from 2024-08-26 14-23-41](https://github.com/user-attachments/assets/03b61bc3-9ea7-4824-be85-f070de7ba3d4)

If you can switch to AMDVLK and use that, it should work fine.

GamerzHell9137 commented:
Fixes missing body parts (arms, eyes) in Kirby's Dream Buffef on RTX 30 series.

Before:

![image-1](https://github.com/user-attachments/assets/3534f205-96d0-4f61-a7b7-3779ca711a17)

After:

![image](https://github.com/user-attachments/assets/7225704f-9415-4e30-be1b-1c801689886d)

BooBerry commented:
New commit seems to fix it for RADV now. Will test Windows in a little bit, but so far so good.

Thanks a lot for fixing this!

GamerzHell9137 commented:
Fixes the regression in Mario + Rabbids: Sparks of Hope where objects flickered in and out of existence on the RTX 30 series.

Before:

![image](https://github.com/user-attachments/assets/4cbd56a9-6598-4a53-81bf-2721be999c80)

After:
![image](https://github.com/user-attachments/assets/bba1342f-f135-4695-aaa1-b9429cd7b646)

Also someone is reporting that in Mario + Rabbids: Kingdom Battle the UI keeps disappearing and appearing on a RTX 40 series card, the issue is not happening on the 30 series cards with or without the PR.

creativecoder commented:
Confirming that the latest version here fixes the [TOTK visual artifacts with RADV](https://github.com/Ryujinx/Ryujinx/pull/7226#issuecomment-2312809787) I see on linux (Pop!_OS 22.04, Radeon 7800 XT, Mesa 24.0.3) 🎉

riperiperi commented:
I'm putting this up for review for now with some things left to future PRs. I've had to work around a bunch of other stuff:

- RADV doesn't support feedback loops on general layout because they didn't foresee the obvious use case, so I've added some code to disable dcc globally to avoid the issue. That's all the feedback loop layout does, anyways (apart from make the texture ineligible for storage and transfer).
  - Maybe in future we use the layout, but I'm not up for implementing it now.
- Added special Silk.NET loader behaviour for MVK. Can be removed when we use a later version (after occlusion query fixes, I guess) and figure out the Vulkan loader.
- Removed some subpass dependency flags from pipeline to prevent every pipeline from throwing a validation error. The errors still appear when write to read barriers are inserted mid-render pass, which is only allowed on non-TDBR gpus. I don't think it is actually possible to do these mid render pass - the only valid dependency you can use here is in framebuffer space, and it's for tile local barriers for feedback loops. This does raise more questions, such as "why did they even add this when feedback loops were not originally possible".

There's also more we can do with feedback loops. We could detect the case where a target is bound twice and force barriers every draw to partially emulate the attachment ordering that seems to happen on Maxwell, and we could split the render pass every draw to prevent cases where the Adreno driver just seems to give up.

gdkchan commented on src/Ryujinx.Graphics.Vulkan/PipelineBase.cs:
```diff
@@ -1493,19 +1526,93 @@ private void RecreateComputePipelineIfNeeded()
+            {
+                if (Gd.Capabilities.SupportsDynamicAttachmentFeedbackLoop)
+                {
+                    DynamicState.SetFeedbackLoop(
```
It could also pass `FeedbackLoopAspects` directly here, but that would require moving the enum, since it's on `PipelineBase` right now. It could also use `ImageAspectFlags` directly instead of this enum, although it has more flag than needed.

riperiperi commented on src/Ryujinx.Common/GraphicsDriver/DriverUtilities.cs:
```diff
@@ -1,13 +1,33 @@
+        {
+            if (OperatingSystem.IsLinux())
+            {
+                AddMesaFlags("RADV_DEBUG", "nodcc");
```
RADV driver disables dcc when a texture has the feedback loop layout. Disabling it for everything is a workaround for RADV not caring that the general layout + feedback loop usage flag should also be able to do that. The Windows driver disables DCC for all textures with the usage flag _regardless_ of the layout, neither of them check if the active pipeline/render target actually signify use of a feedback loop.

We could switch the layout to technically get better performance on RADV, but the problem is that we might need to constantly switch layout to and from general for operations like transfer and storage, since the feedback loop layout only supports read-only sampling and render pass attachment. If RADV were to change to allow feedback loops with the general layout (but not checking actual usage before disabling DCC), we could recreate the texture with the usage flag only when it is first used in a feedback loop, but it's not consistent enough to do that.

riperiperi replied:
I don't actually know how much performance DCC loses, I'm just guessing it won't really matter for switch games that are expected work with constrained bandwidth anyways.

K0bin replied:
> that the general layout + feedback loop usage flag should also be able to do that.

Where does the spec say that?

All it says about feedback loop usage is:

> VK_IMAGE_USAGE_ATTACHMENT_FEEDBACK_LOOP_BIT_EXT specifies that the image can be transitioned to the VK_IMAGE_LAYOUT_ATTACHMENT_FEEDBACK_LOOP_OPTIMAL_EXT layout

So Ryujinx has to transition problematic textures to the FEEDBACK_LOOP_OPTIMAL layout when it encounters a feedback loop. Setting RADV_DEBUG like that is an ugly hack that works around an emulator bug.

riperiperi replied:
It’s more in the documentation for the general layout than in the feedback layout proposal (which is shockingly sparse and this should have come up in Q&A):

`VK_IMAGE_LAYOUT_GENERAL supports all types of device access`

“All” is a pretty strong word. Other cases of adding an “optimal” layout via an extension all specify that the GENERAL layout is also accepted, such as `VK_IMAGE_LAYOUT_FRAGMENT_DENSITY_MAP_OPTIMAL_EXT`, which also has `VK_IMAGE_USAGE_FRAGMENT_DENSITY_MAP_BIT_EXT` to signify that in the worst case (the general layout) the texture needs to support this feature. The word “optimal” here is used for other layouts that are _supported_ by the general layout, but might be faster in the _optimal_ layouts. You don’t need to use the COLOR_ATTACHMENT_OPTIMAL layout to render to a texture, it just might be faster. On a lot of platforms, it’s not faster at all, so using general layout can save a lot if effort managing layout transitions, recreating render passes and descriptors when the layout changes, etc.

It is explicitly stated in the PRs for supporting this extension that performance in DXVK is a priority, so I’d argue that they’re designing to DXVK’s existing behaviours instead of the spec. The turnip driver actually avoids requiring the layout, but this seems like a side effect of how the driver must avoid feedback loop issues by disabling gmem on the pipeline, rather than a texture property (we have obviously implemented the pipeline flags for feedback loops). DXVK performance was also mentioned in this PR, but this time it did not require the layout…

Performance hits for usage flags are also not uncommon, storage usage flag and format alias flag can have big performance implications even if you don’t actually use those features, so if you really want optimal performance you need to create with a minimal usage set and recreate the texture when the usage arises. I think feedback loop is another one of those cases. 

The official AMD driver simply disables dcc on the texture if the usage flag bit is set. That is a reasonable cause and effect for me, and I can mitigate it if I want by recreating the image if it needs that capability, as I need to do for storage usage.

The cost for storage usage cannot be mitigated by switching to another image layout. It doesn’t even have an image layout, you need to use general.

Ideally we would live in a world where image transitions instead worked with usage bits so we could optionally add usages as they are needed rather than recreating the texture (or implementing image transitions through our entire application, assuming a Storage Compatible layout were ever to exist), but that is not the world we live in.

I think there is a gap in the spec here, but I don’t think it’s on RADV’s side. This would be the exception to the rule posed by other image layouts; I don’t think anything but general layout should ever be _required_ to do something as common as an attachment feedback loop.

Joshua-Ashton replied:
Yes... General does support all types of device accesses, but I think you are misinterpreting what thats means -- it has nothing to do with hazards. It can do those accesses, but it cannot guarantee those are not hazardous.

Feedback loops are only allowed with feedback loop layout and not general. All the spec language around relaxing some of those constraints refers to the feedback loop layout only.

As for your comments about "RADV was designed for DXVK and not the spec": Designing for DXVK and the spec is one in the same. I wrote the spec... and the DXVK implementation, and worked on the Mesa implememtation 😄 

The official AMD driver's behaviour is bad and lazy. They should fix it.

K0bin replied:
VK_EXT_image_compression_control can be used to explicitly disable FCC.

Ideally you'd actually transition layouts instead of keeping everything in GENERAL but if you insist, using that extension would still be a better workaround than setting the RADV_DEBUG environment variable.

BNieuwenhuizen replied:
Note that disabling compression doesn't provide any spec guarantees. 

The spec is very clear that the only 2 ways to legally have a feedback loop are input attachments and `VK_IMAGE_LAYOUT_ATTACHMENT_FEEDBACK_LOOP_OPTIMAL_EXT`. See 8.0.2 in the spec (version 1.3.295):


> When relying on non-overlapping accesses between attachments and other resources, it is important to note that [load](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap8.html#renderpass-load-operations) and [store](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap8.html#renderpass-store-operations) operations have fairly wide alignment requirements - potentially affecting entire subresources and adjacent depth/stencil aspects. This makes it invalid to access a non-attachment subresource that is simultaneously being used as an attachment where either access performs a write operation.
>
> The only exception to this is if a subresource is in the VK_IMAGE_LAYOUT_ATTACHMENT_FEEDBACK_LOOP_OPTIMAL_EXT image layout, in which case the overlap is defined to occur at a per-pixel granularity, and applications can read data from pixels outside the render area without introducing a data race.

Anything else and you're relying on specific driver behavior outside of the Vulkan specification.

riperiperi replied:
I don’t think the way you intend the extension to work counts as hazard tracking if a gpu were to need it, it still just counts as saying that there _might_ be a hazard. The peculiar part of the description of the layout is where it says `and/or`. When the image switches to the feedback loop layout, it’s must be used as an attachment _and/or_ sampled. It doesn’t mean that _all_ uses of the texture in this layout are actively participating in a feedback loop. 

All textures used in a render pass could use the feedback loop layout, but only a few of them might be sampled. All textures being sampled might have the layout, but only some of them might be bound. You can retain the layout without there being a feedback loop present at all. I don’t see this as any different from a usage flag (“might” be used for image storage), or any of the other cases where the general layout has been extended to include new usages (with optimal layouts) via extensions. If you absolutely _need_ to know if a feedback loop is happening right now, you still need to check yourself.

I do see it as being an exception and the fact that it doesn’t count as hazard tracking just makes it stand out more. It should have been discussed.

As for VK_EXT_image_compression_control, it is more effort to implement this extension for the one driver it exists for than to pass this flag to RADV for behaviour I don’t think makes sense. We don’t yet have cases where feedback usage flag is omitted (unless the texture cannot be an attachment), so there would be no functional benefit over the global disable right now. Also, I’d rather implement less driver specific paths and extensions, we already have enough for MoltenVK.

Joshua-Ashton replied:
I've read this over a few times and I have no idea at all what you are trying to say really...

You are right that the layout does not mean there is a hazard, just that there might be and it relaxes some rules to allow it.

The usage flag does not relax those rules from a spec point of view.

I don't get what your actual point is though.

riperiperi replied:
The point is that I don’t think there’s any difference in how the layout indicates that a hazard might be present and how the _usage flags_ do. Meaning, I still don’t think the special image layout should be the only way of making sure a texture works with a feedback loop. From what I can see, the extension doesn’t actually allow us to tell the GPU that a hazard is present, only provide a collection of hints that it might be present:

- Pipeline flags
  - A draw with this pipeline might have a draw on either the color or depthstencil attachment. No particular information on what attachment and components/aspect the feedback loop happens on (which _would_ count as exhaustive hazard tracking if it’s required to be correct, but since it’s not here it’s just a hint)
- Dependency flags
  - Render pass with this flag in a subpass dependency might have a framebuffer scope barrier that works for a feedback loop? There is very little information on what this flag is for.
- Usage flags
  - This image _might_ be used in a feedback loop over its lifetime.
- Image layout
  - This image _might_ be used in a feedback loop (but is more likely to) while the layout is active.
  - Also, you can’t use it for anything other than attachment/sample.

The major point is the final two, and how they specifically apply to the image. They both provide hints that would affect properties on the image preventing feedback loops, very similar to how special usage flags like storage and fragment density map work. It’s not particularly special. I’m arguing that the image layout is unfit for controlling the ability for the texture to be used in a feedback loop, because everywhere else in the spec the usage controls this when the general image layout is active.

I’m not sure what’s going to happen on a GPU where these hints are insufficient, since the attachments that are actually used in the feedback loop aren’t fully specified on either the pipeline or the render pass. (for example, if it requires more than an image layout change). They’d probably have to do their own hazard tracking… so the work is done twice.

It doesn’t fit in with the other layouts, either. All other layouts _limit_ the usage for an image to potentially improve performance for heavy use of a single kind, from the full set in the general layout governed by the usage flags. This layout supposedly _adds_ a usage that’s otherwise impossible for _more than one_ use. I don’t believe image layout is an appropriate way to _add_ a type of usage, because it complicates both the image layout and usage flag’s purpose.

If you really want to situationally avoid the cost of a usage flag, adding some new mechanism for allocating a texture with a superset of usage flags (to make sure it has space for all of those usages) and then being able to transition between a subset of those flags would be a better fit. Otherwise, the application should be reallocating the image as it needs to do for storage usage.

Joshua-Ashton replied:
> The point is that I don’t think there’s any difference in how the layout indicates that a hazard might be present and how the usage flags do. 

There is plenty of spec language that does have that difference. The usage flag just says you can use this layout. The layout says you can use certain feedback loops that are otherwise considered data races in that layout.

The flag does not mean you can use it in hazards without the layout at all. That is just wrong.

> Dependency flags 

The dependency flag is just there to signal that there is a feedback loop within the subpasses. DXVK does not use subpasses so it does not use the flag.

There is some CTS that has an example that can explain it better than I can.

> I’m not sure what’s going to happen on a GPU where these hints are insufficient, since the attachments that are actually used in the feedback loop aren’t fully specified on either the pipeline or the render pass. (for example, if it requires more than an image layout change). They’d probably have to do their own hazard tracking… so the work is done twice.

What is the point of discussing this imaginary GPU that you have deemed the extension to not work for?

It is not possible for drivers to do hazard tracking at all in any meaningful way, due to bindless, bda, disjoint regions, control flow, etc etc.

> If you really want to situationally avoid the cost of a usage flag, adding some new mechanism for allocating a texture with a superset of usage flags (to make sure it has space for all of those usages) and then being able to transition between a subset of those flags would be a better fit. Otherwise, the application should be reallocating the image as it needs to do for storage usage.

This is not possible. Some implementations need to change how they use compression or if they use compression and that depends based on their formats to allow for feedback loops and that lasts for the lifetime of the resource.

riperiperi replied:
> What is the point of discussing this imaginary GPU that you have deemed the extension to not work for?

If we’re implementing 4 different hints and just saying “pick one” I don’t think that’s functionally any different from implementing 4 different driver specific extensions to disable DCC, GMEM etc. 

> This is not possible. Some implementations need to change how they use compression or if they use compression and that depends based on their formats to allow for feedback loops and that lasts for the lifetime of the resource.

I don’t see why not, image layout transitions are supposed to do the same thing. If you need more or less space for the image allocation, allocate the superset of all possible usages. I also have no problem manually reallocating the image, because that is consistent with all other usage types and doesn’t need to be reversed when I want to do literally anything else with the image.

I’ve made my case. It’s really not worth refactoring everything in our bindings, render pass and framebuffer caches to switch layouts on a dime for an image flag that could easily stay set permanebtly, and then _have_ to switch it _back_ to perform transfer and storage. If you’re focused on targetting existing gpus, then it should be noted that transfer usage should still work with DCC disabled, but we don’t get access to that capability and have to transition the layout and update pretty much every structure that could ever reference the image.

K0bin replied:
If you insist on keeping the invalid Vulkan usage, I'd suggest at least using the image_compression_control extension to disable DCC instead of the DEBUG environment variable.

riperiperi replied:
I will only do this when reallocation and copy for image usage change is implemented for feedback loops. Otherwise, there is no functional difference between the extension and debug flag, and implementing the extension isn’t exactly lightweight.

K0bin replied:
FWIW the spec is pretty clear that the layout is necessary and LAYOUT_GENERAL must not be used. 

> If the subresource is in the VK_IMAGE_LAYOUT_ATTACHMENT_FEEDBACK_LOOP_OPTIMAL_EXT layout, implementations must not access pixels outside of the render area.

https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap8.html#renderpass-store-operations

> When relying on non-overlapping accesses between attachments and other resources, it is important to note that [load](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap8.html#renderpass-load-operations) and [store](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap8.html#renderpass-store-operations) operations have fairly wide alignment requirements - potentially affecting entire subresources and adjacent depth/stencil aspects. This makes it invalid to access a non-attachment subresource that is simultaneously being used as an attachment where either access performs a write operation.

> The only exception to this is if a subresource is in the VK_IMAGE_LAYOUT_ATTACHMENT_FEEDBACK_LOOP_OPTIMAL_EXT image layout, in which case the overlap is defined to occur at a per-pixel granularity, and applications can read data from pixels outside the render area without introducing a data race.

https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap8.html#_non_overlapping_access_between_resources

riperiperi replied:
Your first and second quote actually has nothing to do with the issue at hand. The problem with the third is that this _should_ include the general layout, otherwise there is no purpose to the image usage flag. My qualm is mostly with the extension’s fundamental premise and how there are clear issues that were not at all discussed. 

I don’t really care how it is intended to be implemented, the spec needs to be changed so that `VK_IMAGE_LAYOUT_GENERAL supports all types of device access` has a specific exclusion of feedback loops, as right now it’s just completely wrong unless the extension is revised without the layout requirement. If this wording can’t be changed to mention that without raising huge questions, consider changing the extension.

Future emulators and compatibility layers shouldn’t be mislead into building their Vulkan backends under the false pretense that fiddling around with image layouts isn’t required if this is the direction the API is headed.

K0bin replied:
I opened a spec issue. Let's see what happens.

riperiperi commented on src/Ryujinx.Common/Utilities/OsUtils.cs:
```diff
@@ -0,0 +1,24 @@
+        public static void SetEnvironmentVariableNoCaching(string key, string value)
+        {
+            // Set the value in the cached environment variables, too.
+            Environment.SetEnvironmentVariable(key, value);
```
The .NET `Environment` set of environment variables is cached as early as possible on non-windows systems. That means that it won't see any changes that `setenv` makes (skips the cache), so I just set it in the cached version so there's no confusion if it's read back later.

riperiperi commented on Directory.Packages.props:
```diff
@@ -39,9 +39,9 @@
-    <PackageVersion Include="Silk.NET.Vulkan" Version="2.16.0" />
-    <PackageVersion Include="Silk.NET.Vulkan.Extensions.EXT" Version="2.16.0" />
-    <PackageVersion Include="Silk.NET.Vulkan.Extensions.KHR" Version="2.16.0" />
+    <PackageVersion Include="Silk.NET.Vulkan" Version="2.21.0" />
```
It is required, the feedback loop extensions are not available on 2.16.0, and I don't think we can use newer extension packages without updating Silk.NET.

gdkchan replied:
yea, I know this change requires the Silk.NET update, I just meant it would be better to update it separately and then take this change. It looks like there are some new warnings too, about missing in/ref keyboard, which might come from the Silk.NET update? Anyway, if you prefer to update it on this PR, that is fine for me too.

riperiperi commented:
Rebased without Silk.NET changes, moved FeedbackLoopAspects into its own file and pass that around more places.

gdkchan reviewed (approved):
lgtm, thanks for fixing this issue.

Michelito97 commented:
I wonder if at some point you could give support to the PC mouse with its different profiles for each game?

COME ON RYUJINX TEAM, THAT'S THE ONLY THING YOU'RE
![Captura de pantalla 2024-09-04 152744](https://github.com/user-attachments/assets/0e59fbd6-d027-475c-9238-6a597eb12516)
 MISSING SO THAT PEOPLE WHO CONTINUE WITH THE YUZU EMULATOR CLONES FOR PC GO DIRECTLY TO THIS GREAT EMULATOR THAT SURPASSES THEM BY FAR IN ALL SECTIONS...

I'm sure it can't be that difficult for such a talented group…

Michelito97 commented:
I wonder if at some point you could give support to the PC mouse with its different profiles for each game?

COME ON RYUJINX TEAM, THAT'S THE ONLY THING YOU'RE MISSING SO THAT PEOPLE WHO CONTINUE WITH THE YUZU EMULATOR CLONES FOR PC GO DIRECTLY TO THIS GREAT EMULATOR THAT SURPASSES THEM BY FAR IN ALL SECTIONS...

I'm sure it can't be that difficult for such a talented group...

![Captura de pantalla 2024-09-04 152744](https://github.com/user-attachments/assets/0e87b479-43b3-4d1c-8727-93666147ccd9)
